### PR TITLE
[framework] defer image load to next event loop

### LINKED ISSUE #97990: Platform/Network Images will begin loading in the current animation frame on the web
Flutter Web uses XHR to load both platform and asset images. While the actual load happens asynchronously, constructing and sending the XHR is surprisingly expensive and is currently performed in the animation frame. This can potentially delay the frame end and subsequent animation fames if the app is under heavy load - a common circumstance if the image load is triggered by something like a route navigation.

In the following example from the new gallery, we can see the xhr events are responsible for about ~6.5 ms in a single frame

image

On the other hand, if we defer them to a separate event loop (using Future.delayed), then they have no impact on the frame workload:

image

This has the advantage of not blocking animations under heavy load. This _may_ impact Flutter on mobile and desktop as well, but since the initialization of image loading seems less expensive it may not be worth it to further defer image loads.

### PR DESCRIPTION
**Pull request by jonahwilliams**
Fixes https://github.com/flutter/flutter/issues/97990

Currently flutter web will begin loading images during an animation frame. This can be surprisingly expensive and contributes to making janky frames even jankier.

Instead force them to run during the next event loop by awaiting a zero duration future. `await null` would move to the next microtask which happens at the end of the frame workload

### CODE REVIEW

**Comment by jonahwilliams** [comment]
Maybe there is a better way to structure this?

**Comment by jonahwilliams** [comment]
hmm, need to make this work for tests

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 1]
```diff
@@ -47,6 +47,12 @@ typedef FrameCallback = void Function(Duration timeStamp);
 /// task does not return a value.
 typedef TaskCallback<T> = T Function();
 
+/// Signature for [SchedulerBinding.scheduleAsyncTask] callbacks.
+///
+/// The type argument `T` is the task's return value. Consider `void` if the
+/// task does not return a value.
+typedef AsyncTaskCallback<T> = Future<T> Function();
```
Comment (yjbanov) [inline packages/flutter/lib/src/scheduler/binding.dart:54]: FYI, there is already `AsyncValueGetter` over in foundation with the exact same signature: https://api.flutter.dev/flutter/foundation/AsyncValueGetter.html

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 2]
```diff
@@ -59,21 +65,32 @@ typedef TaskCallback<T> = T Function();
 ///  * [defaultSchedulingStrategy], the default [SchedulingStrategy] for [SchedulerBinding.schedulingStrategy].
 typedef SchedulingStrategy = bool Function({ required int priority, required SchedulerBinding scheduler });
 
-class _TaskEntry<T> {
+abstract class _BaseTask {
+  int get priority;
```
Comment (yjbanov) [inline packages/flutter/lib/src/scheduler/binding.dart:69]: Does priority need to be abstract?

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 3]
```diff
@@ -59,21 +65,32 @@ typedef TaskCallback<T> = T Function();
...
+
+  StackTrace get debugStack;
```
Comment (yjbanov) [inline packages/flutter/lib/src/scheduler/binding.dart:71]: Does debugStack need to be abstract?

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 4]
```diff
@@ -47,6 +47,12 @@ typedef FrameCallback = void Function(Duration timeStamp);
 /// task does not return a value.
 typedef TaskCallback<T> = T Function();
 
+/// Signature for [SchedulerBinding.scheduleAsyncTask] callbacks.
+///
+/// The type argument `T` is the task's return value. Consider `void` if the
```
Comment (yjbanov) [inline packages/flutter/lib/src/scheduler/binding.dart:52]: nit: "Consider using `void` ..."

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 5]
```diff
@@ -414,6 +458,28 @@ mixin SchedulerBinding on BindingBase {
     return entry.completer.future;
   }
 
+  /// Schedules the given `task` with the given `priority` and returns a
+  /// [Future] that completes to the `task`'s eventual return value.
+  ///
+  /// Like [scheduleTask] except that it handles async tasks.
+  ///
+  /// See also:
+  ///   * [scheduleTask] for a description of the priority system.
```
Comment (yjbanov) [inline packages/flutter/lib/src/scheduler/binding.dart:467]: nit: dartdoc template can be used to inline the description of the priority system.

File: packages/flutter/lib/src/scheduler/binding.dart [hunk 6]
```diff
@@ -432,13 +498,21 @@ mixin SchedulerBinding on BindingBase {
...
     _hasRequestedAnEventLoopCallback = true;
-    Timer.run(_runTasks);
+    scheduleTasks(_runTasks);
+  }
+
+  /// Schedules tasks outside of the current event loop.
+  ///
+  /// By default this uses [Timer.run].
+  @protected
+  void scheduleTasks(void Function() callback) {
```
Comment (yjbanov) [inline packages/flutter/lib/src/scheduler/binding.dart:508]: The method name and dartdocs refer to "tasks" but it's unclear what those tasks are.

**Comment by goderbauer** [comment]
(Triage) @jonahwilliams Do you still have plans for this one?

**Pull request closed without merging**